### DISCUSSION TITLE
Gamepad Support in HTML5 Builds

### DISCUSSION THREAD

**Discussion by ezacharyk** [HaxeFlixel/flixel#2414] [Q&A] [answered]
I am working on getting a game controller (a XBox 360 Controller currently) working in my game. I am setting up these games to be playable in the browser. And while I have the controller input working in my desktop build, the HTML5 build crashes when it enters my play state. I tried compiling the GamePadTest demo to HTML5 but it fails with this error:

`Warning: Could not generate *.woff web font for "/home/zachary/haxelib/flixel/4,10,0/assets/fonts/nokiafc22.ttf"
Warning: Could not generate *.woff web font for "/home/zachary/haxelib/flixel/4,10,0/assets/fonts/monsterrat.ttf"

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: listen EADDRINUSE
    at errnoException (net.js:904:11)
    at Server._listen2 (net.js:1042:14)
    at listen (net.js:1064:10)
    at net.js:1146:9
    at dns.js:72:18
    at process._tickCallback (node.js:419:13)
    at Function.Module.runMain (module.js:499:11)
    at startup (node.js:119:16)
    at node.js:906:3`

And when I try the running demo on the HaxeFlixel site, it doesn't detect my controller. 

Do HTML5 games just not support controllers or am I missing something?

Thanks

**Comment by Geokureli** [comment, marked as answer]
When you run `haxelib run lime test html5` it runs on a port, by default mine runs on `http://127.0.0.1:3000/`, the error you see means that you tried to test it on a port that's actively used by something else, likely your game. try:
`haxelib run lime test html5 --port=3001`

**Reply by ezacharyk** [reply]
Thanks. That solves the issue of the GamepadTest demo not building. However, it still tells me that no controllers are detected. 

So my main concern about controller support for HTML5 games is still present.

**Reply by Geokureli** [reply]
https://haxeflixel.com/demos/GamepadTest/
This works for me, on HTML5 i have to press buttons after clicking the game for focus. only then will it register the gamepad is active

**Reply by ezacharyk** [reply]
Huh. Just tested it with Chrome and it works. So the issue seems to be Firefox then. I will have to look into this more. 

Thanks for your help.

**Comment by ezacharyk** [comment]
Just a note for people who are reading this. 

I initially was referencing the GamepadTest demo for controller input, but there was just way too much going on in that code and it was not fun to try and parse out what I needed. 

So, I switched over to the Mode demo and it used the FlxAction and FlxActionManager classes for input. That was much easier for me to work with and seems to be the solution that I will be using going forward.

I am still unable to use my controller in Firefox on Linux. Haven't tested any other OS yet. But Chrome and native desktop apps work fine.

**Reply by Geokureli** [reply]
I'm having issues with firefox on Windows where all of the buttons in the GamepadTest demo flicker randomly. Firefox for mac just never finds the Gamepad.

For me Brave works on Windows and Mac, Chrome works on Mac, I haven't tried Chrome on Windows, yet

**Reply by ezacharyk** [reply]
Thanks for providing those test cases. That is good to know.
